src/transports/ws/sws.c

```diff
@@ -679,7 +686,75 @@ static void nn_sws_validate_utf8_chunk (struct nn_sws *self)
     return;
 }
 
-static int nn_sws_fail_conn (struct nn_sws *self, int code, char *reason)
+static void nn_sws_acknowledge_close_handshake (struct nn_sws *self)
+{
+    uint8_t *pos;
+    uint16_t close_code;
+    int code_point_len;
+    int len;
+    
+    len = self->inmsg_current_chunk_len;
+    pos = self->inmsg_current_chunk_buf;
+
+    /*  Peer did not provide a Close Code, so choose our own here. */
+    if (len == 0) {
+        nn_sws_fail_conn (self, NN_SWS_CLOSE_NORMAL, "Farewell. <3 nanomsg");
+        return;
+    }
+
+    /*  If the payload is not even long enough for the required 2-octet
+        Close Code, the connection should have been failed upstream. */
+    nn_assert (len >= NN_SWS_CLOSE_CODE_LEN);
+    len -= NN_SWS_CLOSE_CODE_LEN;
+    pos += NN_SWS_CLOSE_CODE_LEN;
+
+    /*  As per RFC 6455 7.1.6, the Close Reason following the Close Code
+        must be well-formed UTF-8. */
+    while (len > 0) {
+        code_point_len = nn_utf8_code_point (pos, len);
+
+        if (code_point_len > 0) {
+            len -= code_point_len;
+            pos += code_point_len;
+            continue;
+        }
+        else {
+            /*  RFC 6455 7.1.6 */
+            nn_sws_fail_conn (self, NN_SWS_CLOSE_ERR_PROTO,
+                "Invalid UTF-8 sent as Close Reason.");
+            return;
+        }
+    }
+
+    /*  Entire Close Reason is well-formed UTF-8 (or empty) */
+    nn_assert (len == 0);
+
+    close_code = nn_gets (self->inmsg_current_chunk_buf);
+    
+    if (close_code == NN_SWS_CLOSE_NORMAL ||
+        close_code == NN_SWS_CLOSE_GOING_AWAY ||
+        close_code == NN_SWS_CLOSE_ERR_PROTO ||
+        close_code == NN_SWS_CLOSE_ERR_WUT ||
+        close_code == NN_SWS_CLOSE_ERR_INVALID_FRAME ||
+        close_code == NN_SWS_CLOSE_ERR_POLICY ||
+        close_code == NN_SWS_CLOSE_ERR_TOOBIG ||
+        close_code == NN_SWS_CLOSE_ERR_EXTENSION ||
+        close_code == NN_SWS_CLOSE_ERR_SERVER ||
+        (close_code >= 3000 && close_code <= 3999) ||
+        (close_code >= 4000 && close_code <= 4999)) {
+        /*  Repeat close code, per RFC 6455 7.4.1 and 7.4.2 */
+        nn_sws_fail_conn (self, (int) close_code,
```

Comment:
Add to this.  I'm gonna squash them all into one anyway. Perhaps against multiple issues.